### PR TITLE
Implement a stub pthreads library for `THREAD_MODEL=single`

### LINKED ISSUE #501: Consider including a pthread.h for all targets
As part of the WASIp2 work, I changed the Makefile to produce a separate `include` directory for each target, only adding a pthread.h to the `wasm32-wasi-threads` one (and its `wasm32-wasip1-threads` alias).  That was a change from the old behavior, where we had a single `include` directory for all targets, with the result that they all had pthread.h.

Lately, I've been helping with the WASI port of .NET, where pthread.h is used in the runtime library for all `TARGET_UNIX` platforms.  That worked fine for older versions of wasi-sdk, since pthread.h was (accidentally?) available for the `wasm32-wasi` target, but doesn't work for wasi-sdk 22.

We can certainly change the `#ifdef TARGET_UNIX` conditionals that guard use of pthread.h to e.g. `#if (defined TARGET_UNIX) && !(defined TARGET_WASI)`, but I'm wondering if it would be useful to include some form of pthread.h in the `include` directories for all targets in wasi-libc.  This would make it easier for projects like .NET to upgrade the wasi-sdk version they're using, as well as make it easier to port other software that uses pthread.h but doesn't necessarily ever create more than one thread.

Thoughts?

### PR DESCRIPTION
**Pull request by ArcaneNibble**
~~This patch series first starts with a number of commits stubbing out functions in the existing `THREAD_model=posix` code. According to "The Open Group Base Specifications Issue 7, 2018 edition", there are a number of mandatory functions which have not been provided. There are also some optional functions that have been partially provided in a not-useful way (e.g. get but no set function). For these, I have chosen to clean them up and remove the get functions for consistency.~~ EDIT: These have been split off into separate PRs and merged.

The remainder of the patches then build up a stub implementation of pthreads for `THREAD_MODEL=single`. I have done my best to try to make sure that all functions are as conforming as possible (under the assumption that another thread cannot ever be launched). This means that objects such as mutexes and rwlocks actually do update their state and will correctly fail when locks cannot be acquired.

When an inevitable deadlock occurs, I have chosen to return EDEADLK when it has been explicitly listed as a permissible return value, and to invoke `__builtin_trap` otherwise.

I have tested this by rebuilding libc++ with threads enabled and then smoke-testing Clang/LLVM-on-WASI to make sure that it can compile a simple program. I have not run any more-extensive conformance testing.

Fixes #501

### CODE REVIEW

**Comment by yamt** [comment]
does this break pthread detection mechanisms like cmake find_package(Threads)?

**Comment by ArcaneNibble** [comment]
> does this break pthread detection mechanisms like cmake find_package(Threads)?

I don't know. I don't use cmake beyond building other software that uses it.

Several of the comments in the linked issue are implying that, even if it might, not having these stubs is causing even more difficulty with porting software.

File: libc-top-half/musl/arch/wasm32/pthread_arch.h
Comment (sbc100) [inline libc-top-half/musl/arch/wasm32/pthread_arch.h:5]: Is this needed?  IIUC, TLS stuff just gets lowered away to normal storage when threads are disabled.

File: Makefile
Comment (sbc100) [inline Makefile:369]: There are only two thread models so this can just be `else`

File: Makefile
Comment (sbc100) [inline Makefile:413]: All the the files that are shared between both models can just go in the common section I think.

File: Makefile
Comment (sbc100) [inline Makefile:350]: Perhaps making these additions to the multi-threaded build should be done as its own separate PR?

File: libc-top-half/musl/src/thread/pthread_atfork.c
Comment (sbc100) [inline libc-top-half/musl/src/thread/pthread_atfork.c:57]: When the entire file is different like this perhaps it makes sense to instead use a new file in the stubs directory?  Or in the bottom half?

File: libc-top-half/musl/src/thread/pthread_attr_setguardsize.c
Comment (sbc100) [inline libc-top-half/musl/src/thread/pthread_attr_setguardsize.c:6]: This might warrant a comment?

File: Makefile
Comment (ArcaneNibble) [reply Makefile:350]: Started breaking this up into smaller commits

File: Makefile
Comment (ArcaneNibble) [reply Makefile:369]: Do we _want_ to bake that in as an assumption? Or do we perhaps want to do things like having three thread models: `posix`/`stub`/`single`?

Comment (sbc100) [reply Makefile:369]: I think we we should make this assumption up until the point where its not longer true.  For now there are only two thread models, and I think its fine for the code reflect that.

**Comment by yamt** [comment]
> > does this break pthread detection mechanisms like cmake find_package(Threads)?
> 
> I don't know. I don't use cmake beyond building other software that uses it.
> 
> Several of the comments in the linked issue are implying that, even if it might, not having these stubs is causing even more difficulty with porting software.

having a stub for mutex etc is probably fine.

otoh, i feel it isn't a great idea to provide pthread_create.

**Comment by ArcaneNibble** [comment]
> > > does this break pthread detection mechanisms like cmake find_package(Threads)?
> > 
> > 
> > I don't know. I don't use cmake beyond building other software that uses it.
> > Several of the comments in the linked issue are implying that, even if it might, not having these stubs is causing even more difficulty with porting software.
> 
> having a stub for mutex etc is probably fine.
> 
> otoh, i feel it isn't a great idea to provide pthread_create.

Although there are [also thoughts of potentially changing it](https://github.com/llvm/llvm-project/pull/92677#issuecomment-2227568684), libcxx currently does require/expect pthread_create to exist. Incidentally, it performs detection whether or not pthreads is supported by looking for `pthread_once` and not `pthread_create` (although said detection doesn't actually seem to work properly for some reason, possibly because wasm handles undefined symbols by importing them from env rather than failing?)

**Comment by ArcaneNibble** [comment]
I've rebased this PR and fixed the issues that CI uncovered. I do wish to merge this, but a decision needs to be made about how to handle the "pthreads detection in existing code" issue.

File: libc-top-half/stub-pthreads/spinlock.c
Comment (sbc100) [inline libc-top-half/stub-pthreads/spinlock.c:1]: Should we call this directory `pthread-stubs`?

And maybe it should live at the top level alongside `dlmalloc` and `emmalloc` (which override/replace libc functions too).  @sunfishcode any preference?

**Comment by ArcaneNibble** [comment]
ping?

**Comment by ArcaneNibble** [comment]
The immediate reason is: I want to be able to build clang+LLVM for WASI (using @whitequark's patchset), which requires threading functions inside libcxx. The usage of types such as `std::mutex` is so pervasive throughout the LLVM codebase that it will probably never be acceptable to put compile-time conditionals around all of them. This is what started the desire for _some_ kind of dummy threading stubs.

As far as I can tell, the only possible choices are:
1. do what I've done here, where all of the stubs (i.e. enough to build libcxx with threads enabled) exist inside wasi-libc. Subsequently change wasi-sdk such that all builds have threads enabled inside libcxx. Use the resulting wasi-sdk to build clang+LLVM
2. do choice 1 except introduce three threading models (i.e. "single", "stub", "posix"). This was discussed in a review comment with @sbc100, who didn't seem to see it necessary to do that.
3. make it entirely libcxx's problem
    1. by making libcxx carry stubs for everything (synchronization *and* nonfunctional thread creation). I predict that this would result in simply having this exact same discussion in a brand-new GitHub thread.
    2. by making libcxx carry stubs for synchronization primitives only, still removing e.g. `std::thread`. This would subsequently require hunting down all uses of these types inside the clang+LLVM codebase in order to build the final intended application, but this looks much more doable compared to trying to remove e.g. `std::mutex`. There does exist [some desire to do this](https://github.com/llvm/llvm-project/pull/92677#issuecomment-2227568684). However, this doesn't handle any C code that would need pthreads stubs, only C++ code. I am currently trying to avoid having to spend the effort required to make this option work.
4. reject upstreaming a stub for `pthread_create` here (keeping only synchronization primitive stubs). This would mean that libcxx still wouldn't build.
    1. Fix libcxx to support "synchronization primitives exist but `pthread_create` doesn't" as a platform. I don't know about the libcxx maintainers, but I personally would consider this very inelegant and clearly a hack, as such a "platform" would certainly not be POSIX-conformant. This _can_ technically be done by simply conditionally guarding one line of code inside libcxx in order to end up with a libcxx where `std::thread` exists but always fails, but once again "I predict that this would result in simply having this exact same discussion in a brand-new GitHub thread." Alternatively, this can be done by introducing the ability for libcxx to have `std::thread` etc. deleted but `std::mutex` etc. present. This would be just as difficult as doing option 3.ii.
    2. Do option 3.i or 3.ii anyways. This works but results in twice the amount of stubs existing.
    3. Some downstream user creates a stub for `pthread_create` anyways and subsequently rebuilds libcxx against it. This is strictly worse as you're now doing the exact same thing except for each downstream user.

The tl;dr is "I need libcxx. I am trying to avoid having to put in all of the effort needed to enable a new 'libcxx has synchronization primitives but no thread creation' build mode for libcxx. I also generally don't think omitting `pthread_create` is a particularly good idea."

In general, the code I want to get ported to WASI would look something like

```
// in do_work_{foo|bar|baz|qux}.c
if (use_multiple_threads) {
  pthread_create(...);
} else {
  do_work_synchronously();
}
// in main.c
if (check_command_line_flags("-threads"))
  use_multiple_threads = 1;
```

With full pthreads stubs in wasi-libc, this code can either be ported by doing nothing ("just" don't pass the `-threads` command line argument) or by having a single `#ifdef __WASI__` in main.c, instead of multiple `#ifdef`s sprinkled throughout all of the `do_work_*.c` parts of the program.

Can you perhaps elaborate further as to why you oppose having `pthread_create`?

**Comment by whitequark** [comment]
(Thanks for doing this work, @ArcaneNibble)

**Comment by yamt** [comment]
> The immediate reason is: I want to be able to build clang+LLVM for WASI (using @whitequark's patchset), which requires threading functions inside libcxx. The usage of types such as `std::mutex` is so pervasive throughout the LLVM codebase that it will probably never be acceptable to put compile-time conditionals around all of them. This is what started the desire for _some_ kind of dummy threading stubs.

how about having libwasi-emulated-pthread.a to follow the precedents?

> Can you perhaps elaborate further as to why you oppose having `pthread_create`?

i don't want to break pthread-detection mechanisms around there, including cmake FindThreads.

**Comment by ArcaneNibble** [comment]
> how about having libwasi-emulated-pthread.a to follow the precedents?

Oh! I quite like this idea! I would still want libcxx to be built with threading support enabled for both `single` and `posix` though, and I hope to avoid having to hack up its build system to do something special.

> i don't want to break pthread-detection mechanisms around there, including cmake FindThreads.

The deliberate intention was to cause detection mechanisms to now detect "this platform has pthreads." This is why I implemented all of the stub functions s.t. they comply with the POSIX specifications (rather than just returning 0 like some other examples that have been posted in #501 ). `EAGAIN` is explicitly listed as a permissible way for `pthread_create` to fail, so software should be expected to do _something_ in response.

**Can we get community consensus** on the approach of "put single-threaded pthreads stubs into libwasi-emulated-pthread.a but always build libcxx with threading enabled (thus causing a program to have undefined symbols if it uses libcxx threading but doesn't link the .a file)" before I go ahead and implement it?

**Comment by yamt** [comment]
> I would still want libcxx to be built with threading support enabled for both single and posix though

do you mean to make wasi-sdk ship libcxx built with LIBCXX_ENABLE_THREADS=ON?

why?

**Comment by ArcaneNibble** [comment]
> do you mean to make wasi-sdk ship libcxx built with LIBCXX_ENABLE_THREADS=ON?
> 
> why?

Yes, because I specifically want to easily port software which uses C++11 threading. i.e. the exact same reason why I am trying to enable pthreads stubs

**Comment by ArcaneNibble** [comment]
@yamt I feel like we are repeatedly talking past each other. Is there any way we can jump into a real-time discussion (e.g. IRC)?

**Comment by ArcaneNibble** [comment]
ping, is there some way through which we can arrive at community consensus about how this pthreads stub functionality should be named and exposed?

File: libc-top-half/stub-pthreads/stub-pthreads.c
Comment (bjorn3) [inline libc-top-half/stub-pthreads/stub-pthreads.c:18]: (I don't have write permission for this repo)

It feels to me like pthread_create, pthread_detach and pthread_join should be kept in a separate archive just like mmap (libwasi-emulated-mman.a) and the signal handler stubs (libwasi-emulated-signal.a) pthread_once and pthread_exit are fine to keep in the main libc.a given that (unlike pthread_create and co) they are actual implementations behaving as they should according to the POSIX specification, albeit only working correctly on single threaded targets, but they are only used on single threaded targets anyway.

Comment (whitequark) [reply libc-top-half/stub-pthreads/stub-pthreads.c:18]: I would be happy to test a proposal implementing this as a solution. @sbc100 @yamt Do you think we could use that approach?

**Comment by ArcaneNibble** [comment]
As there have been multiple people in support of something like this, I have:

* moved the stub files to a new top-level directory (not inside libc-top-half anymore)
* linked the synchronization primitives into libc
* put pthread_create, pthread_detach, pthread_join into libwasi-emulated-pthread.a

File: stub-pthreads/stub-pthreads-emulated.c
Comment (bjorn3) [inline stub-pthreads/stub-pthreads-emulated.c:1]: For other stubs that are put in separate archives the respective header file puts all definitions behind a `#define` to prevent accidental usage. For example `_WASI_EMULATED_MMAN` for `sys/mman.h`. All current ones error out with a suggestion to set the right define and add the right staticlib when included, but for pthreads.h this wouldn't really be possible as it contains a mix of fully supported functions and those that merely have a stub. Maybe it would be possible to define the stub functions as macros that expand to `#error` when used?

File: stub-pthreads/README.md
Comment (sbc100) [inline stub-pthreads/README.md:7]: Can you wrap these long lines at 80 cols?

**Review by sbc100 — APPROVED** [review]
This approach seems reasonable to me.

File: stub-pthreads/stub-pthreads-emulated.c
Comment (ArcaneNibble) [reply stub-pthreads/stub-pthreads-emulated.c:1]: I just tried implementing this, but it's surprisingly tricky to get right. First of all, macros can't expand to other preprocessor directives, so you can't do what you've described. You can expand to `_Static_assert` instead.

For some reason, this ends up making a mess of `expected/*/predefined-macros.txt`, but that might be acceptable.

In any case, I have indeed implemented this suggestion

File: stub-pthreads/README.md
Comment (ArcaneNibble) [reply stub-pthreads/README.md]: done

**Review by abrown — COMMENTED** [review]
I like this PR and took a look today. I think my main hang-up is the question about when these stubs should (a) fail with a trap, (b) silently do nothing, and (c) work as intended. I thought we'd have a lot of (b) seeing as they're stubs but understand why we occasionally need (a) I guess. I suppose you chose (c) in some cases, like simple constructors, because it was easy enough to add? Or were there some real needs for this in your Clang use case?

File: stub-pthreads/barrier.c
Comment (whitequark) [reply stub-pthreads/barrier.c:23]: Let's look at it again:

- `pthread_barrier_init` creates a barrier in a way appropriate for a single-threaded environment.
- `pthread_barrier_destroy` destroys a barrier in a way appropriate for a single-threaded environment, which in this case happens to do nothing.
-`pthread_barrier_wait` waits on a barrier in a way appropriate for a single-threaded environment, i.e. traps since this would result in a deadlock, except for the 1-thread case where it returns `PTHREAD_BARRIER_SERIAL_THREAD`.

If I didn't miss anything, this is a conformant single-threaded implementation, so there is nothing that "doesn't work" there.

**Comment by whitequark** [comment]
> Or were there some real needs for this in your Clang use case?

Clang does actually need the initialization and destruction of primitives to succeed, this is not just an arbitrary choice. Overall, I like @ArcaneNibble's chosen strategy, which is to produce a conformant single-threaded implementation that traps on deadlock.

File: stub-pthreads/barrier.c
Comment (abrown) [reply stub-pthreads/barrier.c:23]: I see what you mean; let me look at this again.

**Review by abrown — APPROVED** [review]
I get it, looks good to me. It will be interesting to see how this holds up when we add proper pthread tests. From the poking I did, the implementations included here do make sense under the assumption that this is only run in single-threaded WebAssembly.

**Comment by abrown** [comment]
Oh, one last comment: @ArcaneNibble, as I reviewed this, I noticed that the recent #522 modifies the expected files so we can't merge this as-is. Can you rebase or merge that in?

**Comment by ArcaneNibble** [comment]
Rebased

File: libc-top-half/musl/include/pthread.h
Comment (carlocab) [reply libc-top-half/musl/include/pthread.h:94]: Filed #558 for this.

**Pull request merged**
